cloud-init: add NoCloud vendor-data support

AnupamSingh2004 opened the pull request:
# Add cloud-init NoCloud vendor-data support

## What this PR does / why we need it

This PR implements support for **vendor-data** in cloud-init NoCloud datasource for KubeVirt VirtualMachines, as requested in issue #16156.

Vendor-data is a cloud-init feature that allows cloud providers/administrators to provide additional configuration that gets merged with user-provided cloud-init data. This is useful for:
- Setting organization-wide defaults
- Injecting monitoring agents or security configurations
- Pre-configuring networking or storage settings
- Adding vendor-specific customizations without modifying user data

The implementation follows the same pattern as existing `userData` and `networkData` fields.

## Which issue(s) this PR fixes

Fixes #16156

## Changes Made

### API Schema (`staging/src/kubevirt.io/api/core/v1/schema.go`)

Added three new fields to `CloudInitNoCloudSource`:

```go
// VendorData contains NoCloud inline cloud-init vendordata.
// +optional
VendorData string `json:"vendorData,omitempty"`

// VendorDataBase64 contains NoCloud cloud-init vendordata as a base64 encoded string.
// +optional  
VendorDataBase64 string `json:"vendorDataBase64,omitempty"`

// VendorDataSecretRef references a k8s secret that contains NoCloud vendordata.
// +optional
VendorDataSecretRef *corev1.LocalObjectReference `json:"vendorDataSecretRef,omitempty"`
```

### Core Logic (`pkg/cloud-init/cloud-init.go`)

- Added `VendorData` field to `CloudInitData` struct
- Updated `readCloudInitNoCloudSource()` to read vendor-data from inline, base64, or secret
- Updated `GenerateLocalData()` to write `vendor-data` file to the ISO (NoCloud only)
- Updated `resolveNoCloudSecrets()` to handle `VendorDataSecretRef`
- Updated `findCloudInitNoCloudSecretVolume()` to include vendor-data secret

### Volume Rendering (`pkg/virt-controller/services/rendervolumes.go`)

Added volume mount handling for `VendorDataSecretRef` in the virt-launcher pod.

## Testing

### Unit Tests

Added 6 new test cases in `pkg/cloud-init/cloud-init_test.go`:
- Inline vendor data
- Base64-encoded vendor data  
- Vendor data from secret reference
- Bad base64 vendor data (error handling)
- CamelCase secret key resolution

```bash
$ go test ./pkg/cloud-init/... -v -count=1

=== RUN   TestCloudInit
...
Ran 211 of 211 Specs in 1.013 seconds
SUCCESS! -- 211 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestCloudInit (1.01s)
PASS
ok      kubevirt.io/kubevirt/pkg/cloud-init     1.013s
```

### Functional Testing

Deployed the modified KubeVirt to a test cluster and verified vendor-data works end-to-end:

**1. Cluster Setup:**
```bash
$ make cluster-up
$ make cluster-sync
```

**2. KubeVirt Deployment Status:**
```bash
$ ./kubevirtci/cluster-up/kubectl.sh get pods -n kubevirt
NAME                               READY   STATUS    RESTARTS   AGE
disks-images-provider-v9n6b        1/1     Running   0          14m
virt-api-896cb86d8-n88xc           1/1     Running   0          12m
virt-controller-6955584fd9-4b6g7   1/1     Running   0          12m
virt-controller-6955584fd9-4q4l5   1/1     Running   0          12m
virt-handler-8ftbj                 1/1     Running   0          12m
virt-operator-78b6c8944-7q22v      1/1     Running   0          14m
virt-operator-78b6c8944-qfqj6      1/1     Running   0          14m

$ ./kubevirtci/cluster-up/kubectl.sh -n kubevirt get kv kubevirt
NAME       AGE   PHASE
kubevirt   13m   Deployed
```

**3. Test VM with vendor-data:**

Created a test VM with inline vendor-data:
```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: test-vendordata
spec:
  running: true
  template:
    spec:
      domain:
        devices:
          disks:
            - name: containerdisk
              disk:
                bus: virtio
            - name: cloudinitdisk
              disk:
                bus: virtio
        resources:
          requests:
            memory: 512Mi
      volumes:
        - name: containerdisk
          containerDisk:
            image: registry:5000/kubevirt/fedora-with-test-tooling-container-disk:devel
        - name: cloudinitdisk
          cloudInitNoCloud:
            vendorData: |
              #cloud-config
              write_files:
                - content: |
                    VENDOR_NAME="kubevirt-vendor"
                    VENDOR_VERSION="1.0"
                    PROVISIONED_BY="cloud-init-vendor-data"
                  path: /etc/vendor_specific_data
                  permissions: '0644'
            userData: |
              #cloud-config
              user: fedora
              password: fedora
              chpasswd:
                expire: false
```

**4. VM Running Successfully:**
```bash
$ ./kubevirtci/cluster-up/kubectl.sh get vmi test-vendordata
NAME              AGE     PHASE     IP            NODENAME   READY
test-vendordata   6m13s   Running   10.244.0.30   node01     True
```

**5. Verified cloud-init ISO contains vendor-data:**
```bash
$ ./kubevirtci/cluster-up/kubectl.sh cp default/virt-launcher-test-vendordata-dhmms:/var/run/kubevirt-ephemeral-disks/cloud-init-data/default/test-vendordata/noCloud.iso /tmp/noCloud.iso -c compute

$ sudo mount -o loop /tmp/noCloud.iso /tmp/iso_mount
$ ls -la /tmp/iso_mount
-rw-------  89 107  meta-data
-rw------- 128 107  user-data
-rw------- 211 107  vendor-data   # <-- NEW FILE

$ sudo cat /tmp/iso_mount/vendor-data
#cloud-config
write_files:
  - content: |
      VENDOR_NAME="kubevirt-vendor"
      VENDOR_VERSION="1.0"
      PROVISIONED_BY="cloud-init-vendor-data"
    path: /etc/vendor_specific_data
    permissions: '0644'
```

## Files Changed

```
api/openapi-spec/swagger.json
pkg/cloud-init/cloud-init.go
pkg/cloud-init/cloud-init_test.go
pkg/virt-controller/services/rendervolumes.go
pkg/virt-operator/resource/generate/components/validations_generated.go
staging/src/kubevirt.io/api/apitesting/testdata/HEAD/kubevirt.io.v1.VirtualMachine.json
staging/src/kubevirt.io/api/apitesting/testdata/HEAD/kubevirt.io.v1.VirtualMachine.yaml
staging/src/kubevirt.io/api/apitesting/testdata/HEAD/kubevirt.io.v1.VirtualMachineInstance.json
staging/src/kubevirt.io/api/apitesting/testdata/HEAD/kubevirt.io.v1.VirtualMachineInstance.yaml
staging/src/kubevirt.io/api/core/v1/deepcopy_generated.go
staging/src/kubevirt.io/api/core/v1/schema.go
staging/src/kubevirt.io/api/core/v1/schema_swagger_generated.go
staging/src/kubevirt.io/client-go/api/openapi_generated.go
```

## Usage Example

```yaml
# Option 1: Inline vendor data
cloudInitNoCloud:
  vendorData: |
    #cloud-config
    write_files:
      - content: "vendor config"
        path: /etc/vendor.conf

# Option 2: Base64-encoded vendor data
cloudInitNoCloud:
  vendorDataBase64: I2Nsb3VkLWNvbmZpZwp3cml0ZV9maWxlczoKICAtIGNvbnRlbnQ6ICJ2ZW5kb3IgY29uZmlnIgogICAgcGF0aDogL2V0Yy92ZW5kb3IuY29uZg==

# Option 3: Vendor data from secret
cloudInitNoCloud:
  vendorDataSecretRef:
    name: my-vendor-data-secret
```


```release-note
Add vendor-data support for cloud-init NoCloud datasource with vendorData, vendorDataBase64, and vendorDataSecretRef fields
```

AnupamSingh2004 committed:
cloud-init: add NoCloud vendor-data support

Add vendorData, vendorDataBase64, and vendorDataSecretRef fields to CloudInitNoCloudSource, enabling vendor-specific cloud-init configuration.

Fixes: https://github.com/kubevirt/kubevirt/issues/16156
Signed-off-by: AnupamSingh2004 <sanupam2004@gmail.com>

0xFelix reviewed (commented):
Thanks for picking this up. Since it changes the API, can you create a VEP for it first? It would be a good time to do this now.

0xFelix commented:
Thanks for creating the VEP. Let's see if we can update to so it follows VEP100 more closely. However, I'm unsure if we can still make it in time for 1.8. Probably we should defer this to 1.9.

AnupamSingh2004 committed:
cloud-init: Add vendor-data support for ConfigDrive

Signed-off-by: AnupamSingh2004 <sanupam2004@gmail.com>

AnupamSingh2004 commented:
Thank you for the feedback @0xFelix.

I have implemented vendor-data support for ConfigDrive as requested. The changes include:-

API Changes:

Added [vendorData](vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/code/electron-browser/workbench/workbench.html), vendorDataBase64, and vendorDataSecretRef fields to CloudInitConfigDriveSource
Implementation:

Updated [cloud-init.go](vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) to handle ConfigDrive vendor-data resolution and ISO generation
Updated [rendervolumes.go](vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) to mount [VendorDataSecretRef](vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) secrets for ConfigDrive volumes

Testing:

Added unit tests for ConfigDrive vendor-data functionality
All existing tests pass
I will also update the VEP document to reflect ConfigDrive support.

Please review at your convenience
Thankyou
